src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+	}
+
+	auto size() const { return tiles.size(); }
+
+	auto begin() const { return tiles.begin(); }
+	auto end() const { return tiles.end(); }
+};
+
+static UniqueTiles dedupTiles(Png const &png, DefaultInitVec<AttrmapEntry> &attrmap,
+                              std::vector<Palette> const &palettes,
+                              DefaultInitVec<size_t> const &mappings) {
+	// Iterate throughout the image, generating tile data as we go
+	// (We don't need the full tile data to be able to dedup tiles, but we don't lose anything
+	// by caching the full tile data anyway, so we might as well.)
+	UniqueTiles tiles;
+
+	auto iter = attrmap.begin();
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		auto [tileID, matchType] = tiles.addTile(tile, palettes[mappings[iter->protoPaletteID]]);
+
+		iter->xFlip = matchType == TileData::HFLIP || matchType == TileData::VHFLIP;
+		iter->yFlip = matchType == TileData::VFLIP || matchType == TileData::VHFLIP;
+		assert(tileID < 1 << 10);
+		iter->tileID = tileID;
+
+		++iter;
+	}
+	assert(iter == attrmap.end());
+
+	return tiles; // Copy elision should prevent the contained `unordered_set` from being
+	              // re-constructed
+}
+
+static void outputTileData(UniqueTiles const &tiles) {
+	std::filebuf output;
+	output.open(options.output, std::ios_base::out | std::ios_base::binary);
+
+	uint16_t tileID = 0;
+	for (TileData const *tile : tiles) {
+		assert(tile->tileID == tileID);
+		++tileID;
+		output.sputn(reinterpret_cast<char const *>(tile->data().data()), tile->data().size());
+	}
+}
+
+static void outputTilemap(DefaultInitVec<AttrmapEntry> const &attrmap) {
+	std::filebuf output;
+	output.open(options.tilemap, std::ios_base::out | std::ios_base::binary);
+
+	assert(options.baseTileIDs[0] == 0 && options.baseTileIDs[1] == 0); // TODO: deal with offset
+	for (AttrmapEntry const &entry : attrmap) {
+		output.sputc(entry.tileID & 0xFF);
+	}
+}
+
+static void outputAttrmap(DefaultInitVec<AttrmapEntry> const &attrmap,
+                          DefaultInitVec<size_t> const &mappings) {
+	std::filebuf output;
+	output.open(options.attrmap, std::ios_base::out | std::ios_base::binary);
+
+	assert(options.baseTileIDs[0] == 0 && options.baseTileIDs[1] == 0); // TODO: deal with offset
+	for (AttrmapEntry const &entry : attrmap) {
+		uint8_t attr = entry.xFlip << 5 | entry.yFlip << 6;
+		attr |= (entry.tileID >= options.maxNbTiles[0]) << 3;
+		attr |= mappings[entry.protoPaletteID] & 7;
+		output.sputc(attr);
+	}
+}
+
+} // namespace optimized
+
+void process() {
+	options.verbosePrint("Using libpng %s\n", png_get_libpng_ver(nullptr));
+
+	options.verbosePrint("Reading tiles...\n");
+	Png png(options.input);
+	ImagePalette const &colors = png.getColors();
+
+	// Now, we have all the image's colors in `colors`
+	// The next step is to order the palette
+
+	if (options.beVerbose) {
+		options.verbosePrint("Image colors: [ ");
+		size_t i = 0;
+		for (auto const &slot : colors) {
```

Comment:
`ranges-v3` would be nice here too in 2023: `for (auto &[i, slot] : view::zip(view::indices, colors)) { ... }`. :/ (Basically Python's `enumerate`.)

Reply:
...or Rust's! :P